src/mds/MetricAggregator.cc

```diff
@@ -53,7 +55,9 @@ int MetricAggregator::init() {
 
 void MetricAggregator::shutdown() {
   dout(10) << dendl;
+  ceph_assert(ceph_mutex_is_locked_by_me(mds->mds_lock));
 
+  mds->mds_lock.unlock();
```

Comment:
Rather than do all this with the real mds_lock, why not just read out the address of mds 0 in notify_mdsmap and send the message using a non-rank-based interface? That seems a lot cleaner with the layers and makes locking simpler.

Reply:
IIRC, Patrick and I discussed somewhat this a while back where a message could be sent to an mds without acquiring mds_lock. I think the point where we decided to go ahead with the current implementation was somewhat related to sending mdsmap before sending out the actual message. Thinking again on this -- does it really matter to send an (updated) mdsmap in this scenario?

Reply:
Ohhhh is it trying to make sure the receiving MDS always has a current MDSMap before dealing with requests?
Hmm, that is a thing we don't want to break in the general case, but it looks like these messages already handle being delayed or misdirected. You'd have to make sure it doesn't erroneously start thinking it owns a metrics gathering server, and that if an MDS sends an update to the old mds0 it correctly sends all the same updates to the new one — but since these area all "total" rather than "since last update" values I think it should be fine?

Reply:
right -- guess it doesn't matter if we sent some updates to an old active mds. this would simplify the locking bits.